lib/mirage.ml

```diff
@@ -917,6 +917,67 @@ let ethernet = Type ETHERNET
 let etif network =
   impl ethernet network (module Ethif)
 
+module Arpv4 = struct
+  type t = {
+    ethernet: ethernet impl;
+    clock: clock impl;
+    time: time impl;
+  }
+
+  let name t = 
+    Name.of_key ("arpv4" ^ Impl.name t.ethernet) ~base:"arpv4"
+
+  let module_name t =
+    String.capitalize (name t)
+
+  let packages t =
+    "tcpip" :: Impl.packages t.time @ Impl.packages t.clock @ Impl.packages t.ethernet
+
+  let libraries t =
+    let arp_library = 
+      match !mode with
+      | `Unix | `MacOSX -> "tcpip.arpv4-unix"
+      | `Xen  -> "tcpip.arpv4"
+    in
+    arp_library :: 
+    Impl.libraries t.ethernet @ 
+    Impl.libraries t.clock @
+    Impl.libraries t.ethernet
+
+  let configure t =
+    let name = name t in
+    let mname = module_name t in
+    append_main "module %s = Arpv4.Make(%s)(%s)(%s)" mname
+      (Impl.module_name t.ethernet) 
+      (Impl.module_name t.clock)  
+      (Impl.module_name t.time);
+    newline_main ();
+    append_main "let %s () =" name;
+    append_main "   %s () >>= function" (Impl.name t.ethernet);
+    append_main "   | `Error _ -> %s" (driver_initialisation_error name);
+    append_main "   | `Ok eth  ->";
+    append_main "   %s.connect eth >>= function" mname;
+    append_main "   | `Error _ -> %s" (driver_initialisation_error "ARP");
```

Comment:
there's a good opportunity here to improve the current situation by using `string_error` (or whatever here the name of the error printing function in `V1.DEVICE`) to pretty-print the error in case the configuration failed.